x-pack/filebeat/input/lumberjack/server_test.go

```diff
@@ -0,0 +1,238 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package lumberjack
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/transport/tlscommon"
+	client "github.com/elastic/go-lumber/client/v2"
+)
+
+const testTimeout = 10 * time.Second
+
+func TestServer(t *testing.T) {
+	makeTestConfig := func() config {
+		var c config
+		c.InitDefaults()
+		c.ListenAddress = "localhost:0"
+		c.MaxConnections = 1
+		c.Keepalive = time.Second
+		c.Timeout = time.Second
+		return c
+	}
+
+	t.Run("empty_batch", func(t *testing.T) {
+		testSendReceive(t, makeTestConfig(), 0, nil)
+	})
+
+	t.Run("no tls", func(t *testing.T) {
+		testSendReceive(t, makeTestConfig(), 10, nil)
+	})
+
+	t.Run("tls", func(t *testing.T) {
+		clientConf, serverConf := tlsSetup(t)
+		clientConf.Certificates = nil
+
+		c := makeTestConfig()
+		c.TLS = serverConf
+		c.TLS.ClientAuth = 0 // tls.NoClientCert
+		c.TLS.VerificationMode = tlscommon.VerifyNone
+
+		testSendReceive(t, c, 10, clientConf)
+	})
+
+	t.Run("mutual tls", func(t *testing.T) {
+		clientConf, serverConf := tlsSetup(t)
+
+		c := makeTestConfig()
+		c.TLS = serverConf
+
+		testSendReceive(t, c, 10, clientConf)
+	})
+}
+
+func testSendReceive(t testing.TB, c config, numberOfEvents int, clientTLSConfig *tls.Config) {
+	require.NoError(t, logp.TestingSetup())
+	log := logp.NewLogger(inputName).With("test_name", t.Name())
+
+	ctx, shutdown := context.WithTimeout(context.Background(), testTimeout)
+	t.Cleanup(shutdown)
+	collect := newEventCollector(ctx, numberOfEvents)
+
+	// Start server.
+	s, err := newServer(c, log, collect.Publish, nil)
+	require.NoError(t, err)
+	go func() {
+		<-ctx.Done()
+		s.Close()
+	}()
+
+	// Asynchronously send and receive events.
+	var wg errgroup.Group
+	wg.Go(s.Run)
+	wg.Go(func() error {
+		// The client returns on error or after an E2E ACK is received.
+		// In both cases the test should shutdown.
+		defer shutdown()
+
+		return sendData(ctx, t, s.bindAddress, numberOfEvents, clientTLSConfig)
+	})
+
+	// Wait for the expected number of events.
+	collect.Await(t)
+
+	// Check for errors from client and server.
+	require.NoError(t, wg.Wait())
+}
+
+func sendData(ctx context.Context, t testing.TB, bindAddress string, numberOfEvents int, clientTLSConfig *tls.Config) error {
+	_, port, err := net.SplitHostPort(bindAddress)
+	if err != nil {
+		return err
+	}
+
+	dialFunc := net.Dial
+	if clientTLSConfig != nil {
+		dialer := &tls.Dialer{
+			Config: clientTLSConfig,
+		}
+		dialFunc = dialer.Dial
+	}
+
+	c, err := client.SyncDialWith(dialFunc, net.JoinHostPort("localhost", port))
+	if err != nil {
+		return fmt.Errorf("client dial error: %w", err)
```

Comment:
It's better to use `t.Fatalf` here because it will interrupt the test. There is no point in trying to use `c` if there was an error, also `c` will be `nil`, which will cause panics. 
```suggestion
		return fmt.Fatalf("client dial error: %w", err)
```

Reply:
I see it now, I probably missed it was running from a different goroutine.